File: docs-examples/example-java/src/test/java/io/micronaut/nats/docs/serdes/ProductInfoSerDes.java
package io.micronaut.nats.docs.serdes;

import io.micronaut.context.annotation.Requires;
import io.micronaut.core.convert.ConversionService;
import io.micronaut.core.type.Argument;
import io.micronaut.nats.serdes.NatsMessageSerDes;
import io.nats.client.Message;
import jakarta.inject.Singleton;

import java.nio.charset.StandardCharsets;
import java.util.Optional;
// end::imports[]

@Requires(property = "spec.name", value = "ProductInfoSerDesSpec")
// tag::clazz[]
@Singleton // <1>
public class ProductInfoSerDes implements NatsMessageSerDes<ProductInfo> { // <2>

    private final ConversionService conversionService;

    public ProductInfoSerDes(ConversionService conversionService) { // <3>
        this.conversionService = conversionService;
    }

    @Override
    public ProductInfo deserialize(Message message, Argument<ProductInfo> argument) { // <4>
        String body = new String(message.getData(), StandardCharsets.UTF_8);
        String[] parts = body.split("\\|");
        if (parts.length == 3) {
            String size = parts[0];
            if (size.equals("null")) {
                size = null;
            }

            Optional<Long> count = conversionService.convert(parts[1], Long.class);
            Optional<Boolean> sealed = conversionService.convert(parts[2], Boolean.class);

            if (count.isPresent() && sealed.isPresent()) {
                return new ProductInfo(size, count.get(), sealed.get());
            }
        }
        return null;
    }

    @Override
    public byte[] serialize(ProductInfo data) { // <5>
        if (data == null) {
            return null;
        }
        return (data.getSize() + "|" + data.getCount() + "|" + data.getSealed()).getBytes(StandardCharsets.UTF_8);
    }

    @Override
    public boolean supports(Argument<ProductInfo> argument) { // <6>
        return argument.getType().isAssignableFrom(ProductInfo.class);
    }
}
// end::clazz[]

File: docs-examples/example-python/src/test/python/micronaut/nats/docs/serdes/ProductInfoSerDes.py
from micronaut.context.annotation import Requires
try:
    # tag::imports[]
    from jakarta.inject import Singleton
    from micronaut.core.convert import ConversionService
    from micronaut.core.type import Argument
    from micronaut.nats.serdes import NatsMessageSerDes
    from java.lang import Boolean, Long
    from .ProductInfo import ProductInfo
    from io.nats.client import Message
    # end::imports[]
except ImportError:  # TODO(python): packages under `io.` other than `io.micronaut` cannot be imported at runtime
    from nats.client import Message


@Requires(property="spec.name", value="ProductInfoSerDesSpec")
# tag::clazz[]
@Singleton  # <1>
class ProductInfoSerDes(NatsMessageSerDes):  # <2>

    def __init__(self, conversion_service: ConversionService):  # <3>
        self.conversion_service = conversion_service

    def deserialize(self, message: Message, argument: Argument) -> ProductInfo | None:  # <4>
        body = bytes(message.getData()).decode("utf-8")
        parts = body.split("|")
        if len(parts) == 3:
            size = parts[0]
            if size == "null":
                size = None

            count = self.conversion_service.convert(parts[1], Long)
            sealed = self.conversion_service.convert(parts[2], Boolean)

            if count.isPresent() and sealed.isPresent():
                return ProductInfo(size, count.get(), sealed.get())
        return None

    # TODO(python): a `bytes | None` return type is coerced with Value.asByte(), so the method is declared to return `bytes`
    def serialize(self, data: ProductInfo) -> bytes:  # <5>
        if data is None:
            return None
        return f"{data.size if data.size is not None else 'null'}|{data.count}|{str(data.sealed).lower()}".encode("utf-8")

    def supports(self, argument: Argument) -> bool:  # <6>
        return argument.getType().isAssignableFrom(ProductInfo)
# end::clazz[]
